src/service/viewport/viewport-binding-ios-embed-sd.js

```diff
@@ -0,0 +1,440 @@
+ * @visibleForTesting
+ */
+export class ViewportBindingIosEmbedShadowRoot_ {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private {!../vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(win);
+
+    const doc = this.win.document;
+    const {documentElement} = doc;
+    documentElement.classList.add('i-amphtml-ios-embed-sd');
+
+    const scroller = htmlFor(doc)`
+      <div id="i-amphtml-scroller">
+        <div id="i-amphtml-body-wrapper">
+          <slot></slot>
+        </div>
+      </div>`;
+
+    /** @private @const {!Element} */
+    this.scroller_ = scroller;
+
+    // Wrapper for the `<body>`.
+    /** @private @const {!Element} */
+    this.wrapper_ = dev().assertElement(scroller.firstElementChild);
+
+    // Notice that the -webkit-overflow-scrolling is set later.
+    setImportantStyles(this.scroller_, {
+      'overflow-x': 'hidden',
+      'overflow-y': 'auto',
+      'position': 'absolute',
+      'top': '0',
+      'left': '0',
+      'right': '0',
+      'bottom': '0',
+      'margin': '0',
+      'width': '100%',
+      'box-sizing': 'border-box',
+      'display': 'block',
+      'padding-top': '0px', // Will be updated for top offset.
+      // The scroller must have a 1px transparent border for two purposes:
+      // (1) to cancel out margin collapse in body's children so that position
+      //     absolute element is positioned correctly.
+      // (2) to offset scroll adjustment to 1 to avoid scroll freeze problem.
+      'border-top': '1px solid transparent',
+    });
+    setImportantStyles(this.wrapper_, {
+      'overflow': 'visible',
+      'position': 'relative',
+    });
+    // Other properties will be copied from the `<body>`.
+
+    /** @private @const {!Observable} */
+    this.scrollObservable_ = new Observable();
+
+    /** @private @const {!Observable} */
+    this.resizeObservable_ = new Observable();
+
+    /** @const {function()} */
+    this.boundScrollEventListener_ = this.onScrolled_.bind(this);
+
+    /** @const {function()} */
+    this.boundResizeEventListener_ = this.onResized_.bind(this);
+
+    /** @private @const {boolean} */
+    this.useLayers_ = isExperimentOn(this.win, 'layers');
+
+    /** @private {number} */
+    this.paddingTop_ = 0;
+
+    /** @private {boolean} */
+    this.bodySyncScheduled_ = false;
+
+    // Setup UI.
+    /** @private {boolean} */
+    this.setupDone_ = false;
+    waitForBody(doc, this.setup_.bind(this));
+
+    // Set overscroll (`-webkit-overflow-scrolling: touch`) later to avoid
+    // iOS rendering bugs. See #8798 for details.
+    whenDocumentReady(doc).then(() => {
+      // There's no way to test `-webkit-overflow-scrolling` style on most of
+      // browsers. Thus we'll use the `i-amphtml-ios-overscroll` class for
+      // testing.
+      this.scroller_.classList.add('i-amphtml-ios-overscroll');
+      setImportantStyles(this.scroller_, {
```

Comment:
Why not add this to the existing rule in main.css with a new selector `#i-amphtml-scroller`?

Reply:
amp.css doesn't reach inside the shadow.

Reply:
Right, derp.